src/logics/aave-v2/service.ts

```diff
@@ -135,80 +73,125 @@ export class Service extends common.Web3Toolkit {
-        return accumulator;
-      }, []);
+      const tokenAddresses: string[] = [];
+      const reserveMap: Record<string, any> = {};
+
+      const lendingPoolAddress = await this.getLendingPoolAddress();
+      const assetAddresses = await LendingPool__factory.connect(lendingPoolAddress, this.provider).getReservesList();
+
+      const calls: common.Multicall3.CallStruct[] = assetAddresses.flatMap((assetAddress) => [
+        {
+          target: this.protocolDataProvider.address,
+          callData: this.protocolDataProviderIface.encodeFunctionData('getReserveConfigurationData', [assetAddress]),
+        },
+        {
+          target: this.protocolDataProvider.address,
+          callData: this.protocolDataProviderIface.encodeFunctionData('getReserveTokensAddresses', [assetAddress]),
+        },
+      ]);
+      const { returnData } = await this.multicall3.callStatic.aggregate(calls);
+
+      assetAddresses.forEach((assetAddress, i) => {
+        const { isActive, isFrozen, borrowingEnabled } = this.protocolDataProviderIface.decodeFunctionResult(
+          'getReserveConfigurationData',
+          returnData[i * 2]
+        );
+
+        reserveMap[assetAddress] = {
+          isSupplyEnabled: isActive && !isFrozen,
+          isBorrowEnabled: isActive && !isFrozen && borrowingEnabled,
+        };
+
+        const { aTokenAddress, stableDebtTokenAddress, variableDebtTokenAddress } =
+          this.protocolDataProviderIface.decodeFunctionResult('getReserveTokensAddresses', returnData[i * 2 + 1]);
+
+        tokenAddresses.push(assetAddress, aTokenAddress, stableDebtTokenAddress, variableDebtTokenAddress);
+      });
+
       const tokens = await this.getTokens(tokenAddresses);
 
-      this.reserveTokens = [];
-      let j = 0;
-      for (let i = 0; i < reserveTokensAddresses.length; i++) {
-        const asset = tokens[j];
-        j++;
-        const aToken = tokens[j];
-        j++;
-        const stableDebtToken = tokens[j];
-        j++;
-        const variableDebtToken = tokens[j];
-        j++;
-        this.reserveTokens.push({ asset, aToken, stableDebtToken, variableDebtToken });
+      for (let i = 0; i < tokens.length; i += 4) {
+        const assetAddress = tokens[i].address;
+
+        reserveMap[assetAddress].asset = tokens[i];
+        reserveMap[assetAddress].aToken = tokens[i + 1];
+        reserveMap[assetAddress].stableDebtToken = tokens[i + 2];
+        reserveMap[assetAddress].variableDebtToken = tokens[i + 3];
       }
+
+      this.reserveTokens = Object.values(reserveMap);
+
+      // Add aToken address as key for quick lookup
+      for (const reserve of Object.values(reserveMap)) {
+        reserveMap[reserve.aToken.address] = reserve;
+      }
+
+      this.reserveMap = reserveMap;
     }
 
-    return this.reserveTokens;
+    return { reserveTokens: this.reserveTokens!, reserveMap: this.reserveMap! };
+  }
+
+  async getSupplyTokens() {
```

Comment:
We made the change to remove the private variable `private assets?: common.Token[]`
because `private reserveTokens?: ReserveTokens[]` is already being stored. The methods `getAssets`, `getATokens`, `getSupplyTokens`, and `getBorrowTokens` are simply performing `map` or `filter` operations, so it might not be necessary to store these tokens separately.